File: src/sdk/python/tests/iotea/core/util/test_json_query.py
##############################################################################
# Copyright (c) 2021 Bosch.IO GmbH
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0
##############################################################################

import pytest
from unittest import TestCase
from src.iotea.core.util.json_query import json_query, json_query_first, json_query_update_first, json_query_update_all

@pytest.fixture
def test_case():
    return TestCase()

@pytest.fixture
def json():
    return {
        'foo': [
            {
                'test': 1
            },
            {
                'test': {
                    'bla': 1234
                }
            },
            {
                'test': 3
            }
        ],
        'bar': {
            'baz': 'Hello World',
            'bla.blubb': 555
        },
        'baz': 42,
        'test': {
            'a': { 'value': 'a' },
            'b': { 'value': 'b' },
            'c': { 'value': 'c' },
            'd': { 'value': 'd' }
        },
        'py': [ 1, 2, 3, 4 ]
    }

class TestJsonQuery:
    def test_should_return_value_identity(self, test_case, json):
        results = json_query(json, '')
        results2 = json_query(json)

        assert len(results) == 1
        assert results[0]['label'] == None
        assert results[0]['query'] == ''
        test_case.assertDictEqual(results[0]['value'], json)
        test_case.assertDictEqual(results[0], results2[0])

    def test_return_label(self, json):
        results = json_query(json, ':foo')

        assert results[0]['label'] == 'foo'

    def test_get_nested_value(self, json):
        results = json_query(json, 'bar.baz')

        assert len(results) == 1
        assert results[0]['value'] == 'Hello World'

    def test_get_all_fields_of_object(self, json):
        results = json_query(json, '*')

        assert len(results) == 5

        for i in range(len(results)):
            assert results[i]['value'] == json[results[i]['query']]

    def test_get_a_field_of_all_nested_objects(self, json):
        results = json_query(json, 'test.*.value')

        assert len(results) == 4

        expected_queries = [
            'test.a.value',
            'test.b.value',
            'test.c.value',
            'test.d.value'
        ]

        for i in range(len(results)):
            expected_queries.index(results[i]['query'])

    def test_get_element_at_given_index(self, json):
        results = json_query(json, 'foo[1].test.bla')

        assert len(results) == 1
        assert results[0]['value'] == 1234

    def test_get_all_array_items(self, json, test_case):
        results = json_query(json, 'foo[:]')

        assert len(results) == 3

        for i in range(len(results)):
            test_case.assertDictEqual(results[i]['value'], json['foo'][i])

    def test_get_specific_array_item(self, json, test_case):
        results = json_query(json, 'foo[1]')

        assert len(results) == 1
        test_case.assertDictEqual(results[0]['value'], json['foo'][1])

    def test_get_last_array_item(self, json, test_case):
        results = json_query(json, 'foo[-1]')
        test_case.assertDictEqual(results[0]['value'], json['foo'][len(json['foo']) - 1])

    def test_get_array_range(self, json, test_case):
        results = json_query(json, 'foo[1:2]')

        assert len(results) == 1
        test_case.assertDictEqual(results[0]['value'], json['foo'][1])

    def test_python_range_selector(self, json, test_case):
        def extract_values(results):
            return list(map(lambda result: result['value'], results))

        results = json_query(json, 'py[1:]')
        test_case.assertListEqual(extract_values(results), [ 2, 3, 4 ])

        results = json_query(json, 'py[:1]')
        test_case.assertListEqual(extract_values(results), [ 1 ])

        results = json_query(json, 'py[3:0]')
        test_case.assertListEqual(extract_values(results), [ ])

        results = json_query(json, 'py[:-1]')
        test_case.assertListEqual(extract_values(results), [ 1, 2, 3 ])

        results = json_query(json, 'py[-3:-1]')
        test_case.assertListEqual(extract_values(results), [ 2, 3 ])

        results = json_query(json, 'py[-1:]')
        test_case.assertListEqual(extract_values(results), [ 4 ])

        results = json_query(json, 'py[0:0]')
        test_case.assertListEqual(extract_values(results), [ ])

        with pytest.raises(Exception) as exc_info:
            json_query(json, 'py[0:5]')

        assert 'Index 5 is out of bounds' == str(exc_info.value)

    def test_accepts_masked_dot_character(self, json, test_case):
        results = json_query(json, "bar.'bla.blubb'")

        assert len(results) == 1
        assert results[0]['value'] == json['bar']['bla.blubb']

    def test_get_first_match(self, json, test_case):
        result = json_query_first(json, 'foo[:]')

        test_case.assertDictEqual(result['value'], json['foo'][0])

    def test_update_first_match(self, json):
        json_query_update_first(json, 'foo[:]', 'huhuhu')

        assert json['foo'][0] == 'huhuhu'

    def test_update_all_matches(self, json, test_case):
        json_query_update_all(json, 'foo[:]', {
            'foo': [
                1, 2, 3
            ]
        })

        test_case.assertEqual(json['foo'], [ 1, 2, 3 ])

File: src/sdk/cpp/lib/tests/test_jsonquery.cpp
/*****************************************************************************
 * Copyright (c) 2021 Bosch.IO GmbH
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 ****************************************************************************/

#include "gtest/gtest.h"
#include "nlohmann/json.hpp"

#include "jsonquery.hpp"

using json = nlohmann::json;

using iotea::core::QueryResult;
using iotea::core::Tokenizer;
using iotea::core::JsonQuery;
using iotea::core::JsonQueryException;

/**
 * @brief Verify that Tokenizer::Next returns the expected tokens.
 */
TEST(jsonquery, Tokenizer_Next) {
    const auto query = "alpha.*.beta[1:2]:label";
    auto t = Tokenizer{query, ".*[]:'"};

    // Verify that all the tokens and nothing but the tokens are returned.
    auto tokens = std::vector<std::string>{ "alpha", ".", "*", ".", "beta", "[", "1", ":", "2", "]", ":", "label" };

    for (const auto& token : tokens) {
        ASSERT_TRUE(t.HasNext());
        ASSERT_EQ(t.Next(), token);
    }
    ASSERT_FALSE(t.HasNext());
}

/**
 * @brief Verify that Tokenizer::PushBack permitts the caller to push the last
 * token back into the stream.
 */
TEST(jsonquery, Tokenizer_PushBack) {
    const auto query = "alpha.*.beta[1:2]:label";
    auto t = Tokenizer{query, ".*[]:'"};

    // Verify that all the tokens and nothing but the tokens are returned.
    auto tokens = std::vector<std::string>{ "alpha", ".", "*", ".", "beta", "[", "1", ":", "2", "]", ":", "label" };

    for (const auto& token : tokens) {
        ASSERT_TRUE(t.HasNext());
        ASSERT_EQ(t.Next(), token);
        t.PushBack();
        ASSERT_TRUE(t.HasNext());
        ASSERT_EQ(t.Next(), token);
    }
    ASSERT_FALSE(t.HasNext());
}

/**
 * @brief Verify that JsonQuery can parse the entire range of query
 * permutations.
 */
TEST(jsonquery, JsonQuery) {
    struct {
        std::string query;
        std::vector<json> value;
        std::vector<std::vector<QueryResult>> want;
    } tests[] {
        {
            "foo.bar:label",
            {
                json::parse(R"({ "foo": { "bar": "baz" } })")
            },
            {
                {
                    QueryResult{"foo.bar", "label", json("baz")}
                }
            }
        },
        {
            "'foo.bar'.baz:label",
            {
                json::parse(R"({ "foo.bar": { "baz": "qux" } })")
            },
            {
                {
                    QueryResult{"'foo.bar'.baz", "label", json("qux")}
                }
            }
        },
        {
            "foo.*:label",
            {
                json::parse(R"({ "foo": { "bar": "baz" } })"),
                json::parse(R"({ "foo": { "bar1": "baz1", "bar2": "baz2" } })")
            },
            {
                {
                    QueryResult{"foo.bar", "label", json("baz")},
                },
                {
                    QueryResult{"foo.bar1", "label", json("baz1")},
                    QueryResult{"foo.bar2", "label", json("baz2")}
                }
            }
        },
        {
            "*.bar:label",
            {
                json::parse(R"({ "foo": { "bar": "baz" } })"),
                json::parse(R"({ "foo1": { "bar": "baz1" }, "foo2": { "bar": "baz2" } })")
            },
            {
                {
                    QueryResult{"foo.bar", "label", json("baz")}
                },
                {
                    QueryResult{"foo1.bar", "label", json("baz1")},
                    QueryResult{"foo2.bar", "label", json("baz2")}
                }
            }
        },
        {
            "*.*:label",
            {
                json::parse(R"({ "foo": { "bar": "baz" } })"),
                json::parse(R"({
                                "foo1": { "bar1": "baz1" },
                                "foo2": { "bar2": "baz2" },
                                "foo3": { "bar3": "baz3" }
                })")
            },
            {
                {
                    QueryResult{"foo.bar", "label", json("baz")}
                },
                {
                    QueryResult{"foo1.bar1", "label", json("baz1")},
                    QueryResult{"foo2.bar2", "label", json("baz2")},
                    QueryResult{"foo3.bar3", "label", json("baz3")}
                }
            }
        },
        {
            "foo[:][1:3]:label",
            {
                json::parse(R"({ "foo": [
                    [1, 2, 3],
                    [4, 5, 6],
                    [7, 8, 9]
                ] })"),
            },
            {
                {
                    QueryResult{"foo[0][1]", "label", json(2)},
                    QueryResult{"foo[0][2]", "label", json(3)},
                    QueryResult{"foo[1][1]", "label", json(5)},
                    QueryResult{"foo[1][2]", "label", json(6)},
                    QueryResult{"foo[2][1]", "label", json(8)},
                    QueryResult{"foo[2][2]", "label", json(9)},
                }
            }
        },
        {
            "foo.bar[:]:label",
            {
                json::parse(R"({ "foo": { "bar": [] } })"),
                json::parse(R"({ "foo": { "bar": [1] } })"),
                json::parse(R"({ "foo": { "bar": [1, 2, 3] } })")
            },
            {
                {
                },
                {
                    QueryResult{"foo.bar[0]", "label", json(1)},
                },
                {
                    QueryResult{"foo.bar[0]", "label", json(1)},
                    QueryResult{"foo.bar[1]", "label", json(2)},
                    QueryResult{"foo.bar[2]", "label", json(3)}
                }
            }
        },
        {
            "foo.bar[0]:label",
            {
                json::parse(R"({ "foo": { "bar": [1] } })"),
                json::parse(R"({ "foo": { "bar": [2, 1] } })"),
            },
            {
                {
                    QueryResult{"foo.bar[0]", "label", json(1)},
                },
                {
                    QueryResult{"foo.bar[0]", "label", json(2)},
                }
            }
        },
        {
            "foo.bar[-1]:label",
            {
                json::parse(R"({ "foo": { "bar": [1] } })"),
                json::parse(R"({ "foo": { "bar": [1, 2] } })"),
            },
            {
                {
                    QueryResult{"foo.bar[0]", "label", json(1)},
                },
                {
                    QueryResult{"foo.bar[1]", "label", json(2)},
                }
            }
        },
        {
            "foo.bar[0:3]:label",
            {
                json::parse(R"({ "foo": { "bar": [1, 2, 3, 4, 5] } })"),
            },
            {
                {
                    QueryResult{"foo.bar[0]", "label", json(1)},
                    QueryResult{"foo.bar[1]", "label", json(2)},
                    QueryResult{"foo.bar[2]", "label", json(3)},
                },
            }
        },
        {
            "foo.bar[2:5]:label",
            {
                json::parse(R"({ "foo": { "bar": [1, 2, 3, 4, 5] } })"),
            },
            {
                {
                    QueryResult{"foo.bar[2]", "label", json(3)},
                    QueryResult{"foo.bar[3]", "label", json(4)},
                    QueryResult{"foo.bar[4]", "label", json(5)},
                },
            }
        },
        {
            "foo.bar.*[1:3].'foo.bar':label",
            {
                json::parse(R"({
                    "foo": {
                        "bar": {
                            "alpha": [
                                {
                                    "foo.bar": 1,
                                    "bar.foo": 2
                                },
                                {
                                    "foo.bar": 3,
                                    "bar.foo": 4
                                },
                                {
                                    "foo.bar": 5,
                                    "bar.foo": 6
                                },
                                {
                                    "foo.bar": 7,
                                    "bar.foo": 8
                                }
                            ],
                            "beta": [
                                {
                                    "foo.bar": 9,
                                    "bar.foo": 10
                                },
                                {
                                    "foo.bar": 11,
                                    "bar.foo": 12
                                },
                                {
                                    "foo.bar": 13,
                                    "bar.foo": 14
                                },
                                {
                                    "foo.bar": 15,
                                    "bar.foo": 16
                                }
                            ]
                        }
                    }
                })")
            },
            {
                {
                    QueryResult{"foo.bar.alpha[1].'foo.bar'", "label", json(3)},
                    QueryResult{"foo.bar.alpha[2].'foo.bar'", "label", json(5)},
                    QueryResult{"foo.bar.beta[1].'foo.bar'", "label", json(11)},
                    QueryResult{"foo.bar.beta[2].'foo.bar'", "label", json(13)}
                }
            }
        }
    };


    for (const auto& t : tests) {
        auto q = JsonQuery{t.query};

        for (size_t i = 0; i < t.value.size(); i++) {
            auto have = q.Query(t.value[i]);
            ASSERT_EQ(have, t.want[i]);
        }
    }
}

/**
 * @brief Verify that JsonQuery throws expections as expected.
 */
TEST(jsonquery, JsonQuery_Exceptions) {
    auto obj = json::parse(R"({"foo": {"bar": [1, 2, 3, 4], "baz": [1, 2, 3, 4]}})");

    // Key not found
    ASSERT_THROW(JsonQuery("foo.car[0]:label").Query(obj), JsonQueryException);

    // Unterminated range
    ASSERT_THROW(JsonQuery("foo.bar[0:label").Query(obj), JsonQueryException);

    // Invalid range parameter
    ASSERT_THROW(JsonQuery("foo.bar[*]:label").Query(obj), JsonQueryException);
    ASSERT_THROW(JsonQuery("foo.bar[0:*]:label").Query(obj), JsonQueryException);
    ASSERT_THROW(JsonQuery("foo.bar[*:0]:label").Query(obj), JsonQueryException);


    // Invalid range (start is after end)
    ASSERT_THROW(JsonQuery("foo.bar[1:0]:label").Query(obj), JsonQueryException);

    // Out of bounds
    ASSERT_THROW(JsonQuery("foo.bar[100]:label").Query(obj), JsonQueryException);
    ASSERT_THROW(JsonQuery("foo.bar[0:100]:label").Query(obj), JsonQueryException);

    // Invalid query (missing label)
    ASSERT_THROW(JsonQuery("foo.bar[0]").Query(obj), JsonQueryException);
}
